docs/about/coding_guidelines.md

```diff
@@ -0,0 +1,68 @@
+make a contribution, by proving a fix to said issue.
+
+Last, but not least, these rules are continuously evolving and as such, 
+refer to them once in a while when in doubt.
+
+
+### Defining classes and traits
+
+1. Value classes must be final and extend AnyVal.
+2. Method extension classes must be final and extend `AnyVal`.
+3. Sealed traits that are ADTs should extend `Product` and `Serializable`.
+4. Regular traits and sealed trait that do not form ADTS should extend `Serializable` but not `Product`.
+5. Multiple parameter classes that are ADT should be case class and not extend AnyVal
+6. Traits should always extends `Serializable`. (i.e. `ZIO`)
+
+### Final and private modifiers 
+
+1. All methods on classes / traits are declared `final`, by default.
+2. No methods on objects declared `final`.
+3. No methods on final classes declared `final`.
+4. All classes inside objects should be defined `final`.
+5. All final classes have private constructors & constructor parameters.
+6. All `vals` declared `final`, either in objects or `final classes`, if they are constant expressions and without type annotations.
+7. Private `methods` and `vals` are not declared `final` since they are already final.
+8. Package-private `vals` and `methods` declared `final`.
+
+### Refactoring
+
+1. If a class has all `final` members, the class should be declared `final` and `final` member annotations should be removed except constant expressions.
+2. All type annotations should use the least powerful type alias. This means, that, let us say, a `ZIO` effect that has no dependencies but throws an arbitrary error, should be defined as `IO`.
+
+### Naming of parameters in methods and classes
+
+1. Partial functions are called `pf`
+2. Evidences are called `ev`
+3. Promise are called `p` (unless in its own class methods, in that case it is called `that`)
+4. Functions are called `fn`, `fn1`, unless they bear specific meaning: `use`, `release`, ..
+4. Consider _ method having more meaningful names
+5. Iterable are called `in`
+6. When a parameter type equals own (in a method of a trait) call it `that`
+7. Never use by-name parameter unless you really know what you are doing and it makes sense. Mind the `Function[0]` extra allocation and loss of clean syntax when invoking the method.
+8. Folds initial value are called `zero`
+
+### Misc
+
+1. GADTS should always have type annotation.
+2. Type alias should always have type annotation.
+3. Methods that lift pure values to effects are dangerous. Dangerous in the sense that they can potentially have dangerous side-effects. 
+Such methods should have a default lazy variant and an eager variant for advanced users that are aware they absolutely do not have side-effects in their code, 
+having slight gains in performance. The lazy variant should have a normal name (succeed, fail, lift, potato) and the eager variant should have a now suffix 
+(succeedNow, liftNow, potatoNow) which makes it clear of its eager behaviour.      
```

Comment:
I like this bit on `succeedNow`, etc., but I think we need a whole section on naming.

It could have things like this in it:

1. Methods that have the form of `List#zip` are called `zip`, and have an alias called `<*>`. The parallel version, if applicable, has the name `zipPar`, with an alias called `<&>`.
2. Methods that are intended to capture side-effects, convert them into functional effects, should be prefixed by `effect*`. For example, `ZIO.effect`.
3. The dual of zip, which is trying _either_ a left or right side, producing an `Either` of the result, should be called `orElseEither`, with alias `<+>`. The simplified variant where both left and right have the same type should be called `orElse`, with alias `<>`.
4. Constructors for a data type `X` that are based on another data type `Y` should be placed in the companion object `X` and named `fromY`. For example, `ZIO.fromOption`, `ZStream.fromEffect`.
5. Parallel versions of methods should be named the same, but with a `Par` suffix. Parallel versions with a bound on parallelism should use a `ParN` suffix.
6. `foreach` should be used as the default traverse operation, with `traverse` retained as an alias for programmers with an FP background. For example, `ZIO.foreach`.


Etc. I can probably think of at least 10 more. But we don't have to get them all into the first PR. 😄

Reply:
This looks great to add!